File: backend/app/schemas/simulation_lab.py
"""Schemas for the BOS Simulation Lab virtual closed-loop harness."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.evidence import SimulationEvidenceSource

SimulationScenario = Literal["normal", "moisture_drift", "temperature_spike", "underfeeding"]
SimulationPolicy = Literal["rule_based", "conservative", "growth_optimized", "risk_minimizing"]


class SimulationInitialState(BaseModel):
    biomass: float = Field(default=0.5, gt=0)
    substrate: float = Field(default=10.0, gt=0)
    temperature: float = Field(default=28.0, ge=0, le=60)
    moisture: float = Field(default=70.0, ge=0, le=100)
    nitrogen: float = Field(default=50.0, ge=0)


class LabAssayMeasurements(BaseModel):
    biomass: float | None = Field(default=None, ge=0)
    substrate: float | None = Field(default=None, ge=0)
    moisture: float | None = Field(default=None, ge=0, le=100)
    heavy_metal_index: float | None = Field(default=None, ge=0, le=1)


class SimulationScenarioCreate(BaseModel):
    species: str = Field(default="BSF", min_length=1, max_length=80)
    feedstock: str = Field(default="mixed_food_waste", min_length=1, max_length=120)
    scenario: SimulationScenario = "normal"
    initial_state: SimulationInitialState = Field(default_factory=SimulationInitialState)
    cycles: int = Field(default=8, ge=5, le=24)
    seed: int = Field(default=7, ge=0, le=1_000_000)
    policy: SimulationPolicy = "rule_based"
    lab_assay_measurements: LabAssayMeasurements | None = None


class SimulationScenarioResponse(BaseModel):
    simulation_id: str
    batch_id: str
    tenant_id: int
    species: str
    feedstock: str
    scenario: SimulationScenario
    initial_state: SimulationInitialState
    cycles: int
    seed: int
    policy: SimulationPolicy
    lab_assay_measurements: LabAssayMeasurements | None = None
    evidence_sources: list[SimulationEvidenceSource] = Field(default_factory=list)
    status: str
    created_at: datetime


class SimulationLabCycleResponse(BaseModel):
    cycle: int
    timestamp: datetime
    state_before: dict[str, Any]
    sensor_observation: dict[str, Any]
    supervisor_decision: dict[str, Any]
    risk_prediction: dict[str, Any]
    visual_observation: dict[str, Any] | None = None
    evidence_sources: list[SimulationEvidenceSource] = Field(default_factory=list)
    agent_action: dict[str, Any]
    actuator_result: dict[str, Any]
    state_after: dict[str, Any]
    audit_event: dict[str, Any]


class SimulationLabSummaryResponse(BaseModel):
    simulation_id: str
    cycle_count: int
    scenario: SimulationScenario
    policy: SimulationPolicy
    starting_risk: float
    ending_risk: float
    risk_delta: float
    action_count: int
    audit_event_count: int
    final_state: dict[str, Any]


class SimulationLabRunResponse(BaseModel):
    scenario: SimulationScenarioResponse
    summary: SimulationLabSummaryResponse
    cycles: list[SimulationLabCycleResponse]


class SimulationLabRunHistoryItem(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    run_id: str
    simulation_id: str
    tenant_id: int
    policy: SimulationPolicy
    status: str
    engine_version: str = "simulation_lab_v2_8"
    model_version: str | None = None
    input_snapshot_id: str | None = None
    input_snapshot_hash: str | None = None
    replay_of_run_id: str | None = None
    evidence_pack_id: str | None = None
    cycle_count: int
    starting_risk: float
    ending_risk: float
    risk_delta: float
    action_count: int
    audit_event_count: int
    final_state: dict[str, Any]
    started_at: datetime
    completed_at: datetime | None = None
    created_at: datetime


class SimulationPolicyCompareRequest(BaseModel):
    policies: list[SimulationPolicy] = Field(
        default_factory=lambda: ["rule_based", "conservative", "growth_optimized", "risk_minimizing"],
        min_length=1,
        max_length=4,
    )
    baseline_policy: SimulationPolicy = "rule_based"


class SimulationScenarioImportRequest(BaseModel):
    reference_id: str = Field(..., min_length=1, max_length=160)
    scenario_name: str | None = Field(default=None, max_length=160)
    policy: SimulationPolicy = "rule_based"
    cycles: int = Field(default=8, ge=5, le=24)
    seed: int = Field(default=17, ge=0, le=1_000_000)


class SimulationScenarioImportMetadata(BaseModel):
    reference_id: str
    reference_source: str
    source_title: str
    review_status: str = "review_required"
    human_review_required: bool = True
    numeric_values_included: bool = True
    release_evidence_allowed: bool = False
    runtime_activation_enabled: bool = False
    validated_default_write_enabled: bool = False
    promotion_enabled: bool = False
    extraction_confidence: float
    field_sources: dict[str, str]
    fallbacks: list[str] = Field(default_factory=list)
    environmental_drift_assumptions: dict[str, Any] = Field(default_factory=dict)
    expected_risk_constraints: dict[str, Any] = Field(default_factory=dict)


class SimulationScenarioImportResponse(BaseModel):
    scenario: SimulationScenarioResponse
    import_metadata: SimulationScenarioImportMetadata


class SimulationPolicyComparisonMetrics(BaseModel):
    risk_delta: float
    biomass_gain: float
    substrate_use: float
    intervention_count: int
    human_review_count: int
    audit_completeness: float
    conversion_rate_estimate: float
    mortality_risk: float
    moisture_risk: float
    nh3_risk: float
    energy_kwh_estimate: float
    water_kg_estimate: float
    co2e_estimate: float
    labor_hour_estimate: float
    gross_margin_estimate: float
    release_readiness: str
    missing_evidence: list[str] = Field(default_factory=list)
    human_review_required: bool = False


class SimulationPolicyComparisonRun(BaseModel):
    policy: SimulationPolicy
    run_id: str
    summary: SimulationLabSummaryResponse
    metrics: SimulationPolicyComparisonMetrics


class SimulationPolicyComparisonResponse(BaseModel):
    simulation_id: str
    baseline_policy: SimulationPolicy
    runs: list[SimulationPolicyComparisonRun]
    winner: dict[str, SimulationPolicy | None]


class SimulationReleaseAppendixAuditHash(BaseModel):
    cycle: int | None = None
    event_type: str
    payload_hash: str
    recorded_at: datetime


class SimulationReleaseAppendixResponse(BaseModel):
    simulation_id: str
    generated_at: datetime
    scenario: SimulationScenarioResponse
    selected_run: SimulationLabRunHistoryItem | None = None
    comparison_runs: list[SimulationPolicyComparisonRun] = Field(default_factory=list)
    audit_trace_hashes: list[SimulationReleaseAppendixAuditHash] = Field(default_factory=list)
    evidence_pack_id: str | None = None
    input_snapshot_id: str | None = None
    disclaimer: str


class SimulationEvidenceSourcesResponse(BaseModel):
    simulation_id: str
    scenario_sources: list[SimulationEvidenceSource] = Field(default_factory=list)
    latest_run_sources: list[SimulationEvidenceSource] = Field(default_factory=list)
    cycle_sources: list[dict[str, Any]] = Field(default_factory=list)


class SimulationReplayResponse(BaseModel):
    source_run_id: str
    replay_run: SimulationLabRunHistoryItem
    deterministic_core_match: bool
    warnings: list[str] = Field(default_factory=list)


class SimulationRunDiffResponse(BaseModel):
    run_id: str
    against_run_id: str
    deterministic_core_match: bool
    differences: list[dict[str, Any]] = Field(default_factory=list)


class SimulationLabCycleListResponse(BaseModel):
    simulation_id: str
    cycles: list[SimulationLabCycleResponse]


class SimulationLabAuditTraceResponse(BaseModel):
    simulation_id: str
    audit_trace: list[dict[str, Any]]


class SimulationLabExportResponse(BaseModel):
    simulation_id: str
    exported_at: datetime
    scenario: SimulationScenarioResponse
    summary: SimulationLabSummaryResponse | None = None
    cycles: list[SimulationLabCycleResponse] = Field(default_factory=list)
    audit_trace: list[dict[str, Any]] = Field(default_factory=list)
    evidence_sources: list[SimulationEvidenceSource] = Field(default_factory=list)
    evidence_pack_id: str | None = None
    input_snapshot_id: str | None = None
    disclaimer: str

File: frontend/src/types/simulationLab.ts
export type SimulationScenario =
  | "normal"
  | "moisture_drift"
  | "temperature_spike"
  | "underfeeding";

export type SimulationPolicy =
  | "rule_based"
  | "conservative"
  | "growth_optimized"
  | "risk_minimizing";

export interface SimulationInitialState {
  biomass: number;
  substrate: number;
  temperature: number;
  moisture: number;
  nitrogen: number;
}

export interface LabAssayMeasurements {
  biomass?: number | null;
  substrate?: number | null;
  moisture?: number | null;
  heavy_metal_index?: number | null;
}

export type EvidenceSourceKind =
  | "operator_input"
  | "imported_reference"
  | "manuscript_campaign"
  | "staged_reference"
  | "promoted_campaign"
  | "synthetic"
  | "deterministic_model"
  | "chronos_model"
  | "fallback_default"
  | "assumed"
  | "official_standard"
  | "peer_reviewed_literature"
  | "public_dataset"
  | "industry_reference"
  | "commercial_database"
  | "model_provider_docs"
  | "github_reference";

export interface SimulationEvidenceSource {
  field: string;
  source_kind: EvidenceSourceKind;
  source_ref: string | null;
  confidence: number;
  fallback_used: boolean;
  notes: string | null;
  payload?: Record<string, number> | null;
  measurement_mode?: string | null;
  review_status?: string | null;
  human_review_required?: boolean | null;
  numeric_values_included?: boolean | null;
  release_evidence_allowed?: boolean | null;
  runtime_activation_enabled?: boolean | null;
  validated_default_write_enabled?: boolean | null;
  promotion_enabled?: boolean | null;
}

export interface SimulationScenarioCreate {
  species: string;
  feedstock: string;
  scenario: SimulationScenario;
  initial_state: SimulationInitialState;
  cycles: number;
  seed: number;
  policy: SimulationPolicy;
  lab_assay_measurements?: LabAssayMeasurements | null;
}

export interface SimulationScenarioResponse extends SimulationScenarioCreate {
  simulation_id: string;
  batch_id: string;
  tenant_id: number;
  status: string;
  created_at: string;
  evidence_sources: SimulationEvidenceSource[];
}

export interface SimulationLabCycle {
  cycle: number;
  timestamp: string;
  state_before: Record<string, number>;
  sensor_observation: Record<string, number | string>;
  supervisor_decision: Record<string, unknown>;
  risk_prediction: {
    future_risk_score: number;
    release_warning_score: number;
    forecast_window: string;
    model_name: string;
    driver_features: Record<string, number>;
    source?: string;
    execution_mode?: string;
    confidence_band?: {
      lower: number;
      median: number;
      upper: number;
    };
    lab_governance?: {
      profile_id: string;
      heavy_metal_gate: string;
      product_use_lock: string;
      review_status: string;
    };
    lab_assay_comparison?: {
      source_kind: string;
      source_ref: string;
      review_status: string;
      measurement_mode: string;
      human_review_required?: boolean;
      numeric_values_included?: boolean;
      release_evidence_allowed?: boolean;
      runtime_activation_enabled?: boolean;
      validated_default_write_enabled?: boolean;
      promotion_enabled?: boolean;
      predicted: Record<string, number>;
      measured_anchor: Record<string, number>;
      residuals: Record<string, number>;
      out_of_band: string[];
      calibration_status: string;
    };
  };
  visual_observation: {
    source: string;
    frame_id: string;
    anomaly_labels: string[];
    confidence: number;
    ultralytics_dry_run: boolean;
    hardware_camera_used: boolean;
  } | null;
  evidence_sources: SimulationEvidenceSource[];
  agent_action: {
    policy: SimulationPolicy;
    recommended_action: string;
    intensity: number;
    duration_minutes: number;
    reason_codes: string[];
    confidence: number;
    expected_risk_reduction: number;
  };
  actuator_result: {
    action: string;
    status: string;
    expected_effect: Record<string, string>;
    actual_effect: Record<string, string>;
    hardware_execution: boolean;
  };
  state_after: Record<string, number>;
  audit_event: Record<string, unknown>;
}

export interface SimulationLabSummary {
  simulation_id: string;
  cycle_count: number;
  scenario: SimulationScenario;
  policy: SimulationPolicy;
  starting_risk: number;
  ending_risk: number;
  risk_delta: number;
  action_count: number;
  audit_event_count: number;
  final_state: Record<string, number>;
}

export interface SimulationLabRunResponse {
  scenario: SimulationScenarioResponse;
  summary: SimulationLabSummary;
  cycles: SimulationLabCycle[];
}

export interface SimulationLabRunHistoryItem extends SimulationLabSummary {
  run_id: string;
  tenant_id: number;
  status: string;
  engine_version: string;
  model_version: string | null;
  input_snapshot_id: string | null;
  input_snapshot_hash: string | null;
  replay_of_run_id: string | null;
  evidence_pack_id: string | null;
  started_at: string;
  completed_at: string | null;
  created_at: string;
}

export interface SimulationPolicyCompareRequest {
  policies: SimulationPolicy[];
  baseline_policy: SimulationPolicy;
}

export interface SimulationScenarioImportRequest {
  reference_id: string;
  scenario_name?: string | null;
  policy: SimulationPolicy;
  cycles: number;
  seed: number;
}

export interface SimulationScenarioImportMetadata {
  reference_id: string;
  reference_source: string;
  source_title: string;
  review_status: string;
  human_review_required: boolean;
  numeric_values_included: boolean;
  release_evidence_allowed: boolean;
  runtime_activation_enabled: boolean;
  validated_default_write_enabled: boolean;
  promotion_enabled: boolean;
  extraction_confidence: number;
  field_sources: Record<string, string>;
  fallbacks: string[];
  environmental_drift_assumptions: Record<string, unknown>;
  expected_risk_constraints: Record<string, unknown>;
}

export interface SimulationScenarioImportResponse {
  scenario: SimulationScenarioResponse;
  import_metadata: SimulationScenarioImportMetadata;
}

export interface SimulationPolicyComparisonMetrics {
  risk_delta: number;
  biomass_gain: number;
  substrate_use: number;
  intervention_count: number;
  human_review_count: number;
  audit_completeness: number;
  conversion_rate_estimate: number;
  mortality_risk: number;
  moisture_risk: number;
  nh3_risk: number;
  energy_kwh_estimate: number;
  water_kg_estimate: number;
  co2e_estimate: number;
  labor_hour_estimate: number;
  gross_margin_estimate: number;
  release_readiness: string;
  missing_evidence: string[];
  human_review_required: boolean;
}

export interface SimulationPolicyComparisonRun {
  policy: SimulationPolicy;
  run_id: string;
  summary: SimulationLabSummary;
  metrics: SimulationPolicyComparisonMetrics;
}

export interface SimulationPolicyComparisonResponse {
  simulation_id: string;
  baseline_policy: SimulationPolicy;
  runs: SimulationPolicyComparisonRun[];
  winner: {
    lowest_risk: SimulationPolicy | null;
    highest_biomass_gain: SimulationPolicy | null;
    fewest_interventions: SimulationPolicy | null;
  };
}

export interface SimulationReleaseAppendixAuditHash {
  cycle: number | null;
  event_type: string;
  payload_hash: string;
  recorded_at: string;
}

export interface SimulationReleaseAppendixResponse {
  simulation_id: string;
  generated_at: string;
  scenario: SimulationScenarioResponse;
  selected_run: SimulationLabRunHistoryItem | null;
  comparison_runs: SimulationPolicyComparisonRun[];
  audit_trace_hashes: SimulationReleaseAppendixAuditHash[];
  evidence_pack_id: string | null;
  input_snapshot_id: string | null;
  disclaimer: string;
}

export interface SimulationEvidenceSourcesResponse {
  simulation_id: string;
  scenario_sources: SimulationEvidenceSource[];
  latest_run_sources: SimulationEvidenceSource[];
  cycle_sources: Array<{ cycle: number; evidence_sources: SimulationEvidenceSource[] }>;
}

export interface SimulationReplayResponse {
  source_run_id: string;
  replay_run: SimulationLabRunHistoryItem;
  deterministic_core_match: boolean;
  warnings: string[];
}

export interface SimulationRunDiffResponse {
  run_id: string;
  against_run_id: string;
  deterministic_core_match: boolean;
  differences: Array<Record<string, unknown>>;
}

export interface SimulationLabCycleListResponse {
  simulation_id: string;
  cycles: SimulationLabCycle[];
}

export interface SimulationLabAuditTraceResponse {
  simulation_id: string;
  audit_trace: Array<Record<string, unknown>>;
}

export interface SimulationLabExportResponse {
  simulation_id: string;
  exported_at: string;
  scenario: SimulationScenarioResponse;
  summary: SimulationLabSummary | null;
  cycles: SimulationLabCycle[];
  audit_trace: Array<Record<string, unknown>>;
  evidence_sources: SimulationEvidenceSource[];
  evidence_pack_id: string | null;
  input_snapshot_id: string | null;
  disclaimer: string;
}
